src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -913,6 +1135,36 @@ func (s *commitLogSource) mostRecentSnapshotByBlockShard(
 	return mostRecentCompleteSnapshotByBlockShard, nil
 }
 
+func (s *commitLogSource) mostRecentIndexSnapshotsByBlock(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	indexSnapshotFiles fs.FileSetFilesSlice,
+	indexInfoFiles []fs.ReadIndexInfoFileResult,
+) (
+	map[xtime.UnixNano]indexSnapshot,
+	error,
+) {
+	blockSize := ns.Options().RetentionOptions().BlockSize()
+
+	mostRecentCompleteIndexSnapshotByBlock := s.mostRecentCompleteIndexSnapshotByBlock(
+		shardsTimeRanges, blockSize, indexSnapshotFiles, indexInfoFiles,
+		s.opts.CommitLogOptions().FilesystemOptions())
+	for block, mostRecent := range mostRecentCompleteIndexSnapshotByBlock {
+		if mostRecent.fileSet.CachedSnapshotTime.IsZero() {
+			// Should never happen.
+			return nil, instrument.InvariantErrorf(
```

Comment:
Can you make this emit a metric as well using `instrument.EmitAndLogInvariantViolation`?

Reply:
After syncing w/ @robskillington offline, decided on scrapping `CachedSnapshotTime` logic entirely and have the data/index snapshot functions only return snapshots for shard and/or block combinations if they exist.